bayesian_logistic_regression.py

```diff
@@ -0,0 +1,365 @@
+            self.initial_loc = torch.zeros(n_variables)
+            self.initial_scale = torch.ones(n_variables)
+        else:
+            self.initial_scale = initial_scale
+            self.initial_loc = initial_loc    
+
+    def get_distribution(self):
+        scale_q = pyro.param('scale_{}'.format(self.index), self.initial_scale, constraints.positive)
+        #scale_q = torch.eye(self.n_variables)
+        locs_q = pyro.param('locs_{}'.format(self.index), self.initial_loc)
+        return dist.MultivariateNormal(locs_q, scale_q).to_event(1)
+
+    def __call__(self, observations, input_data):
+        distribution = self.get_distribution()
+        w = pyro.sample("w", distribution)
+        return w
+        
+def logistic_regression_model(observations, input_data):
+    w = pyro.sample('w', dist.MultivariateNormal(torch.zeros(input_data.shape[1]), torch.ones(input_data.shape[1])).to_event(1))
+    with pyro.plate("data", input_data.shape[0]):
+      sigmoid = torch.sigmoid(torch.matmul(input_data, w.double()))
+      obs = pyro.sample('obs', dist.Bernoulli(sigmoid), obs=observations)
+
+# @config_enumerate
+# def approximation(observations, input_data, components, weights):
+#     assignment = pyro.sample('assignment', dist.Categorical(weights))
+#     distribution = components[assignment].get_distribution()
+#     w = pyro.sample("w", distribution)
+#     return w
+
+def dummy_approximation(observations, input_data):
+    variance_q = pyro.param('variance_0', torch.eye(input_data.shape[1]), constraints.positive)
+    mu_q = pyro.param('mu_0', 100*torch.ones(input_data.shape[1]))
+    pyro.sample("w", dist.MultivariateNormal(mu_q, variance_q))
+
+def predictive_model(wrapped_approximation, observations, input_data):
+    w = wrapped_approximation(observations, input_data)
+    if type(w) is dict:
+        w = w['w']
+    with pyro.plate("data", input_data.shape[0]):
+      sigmoid = torch.sigmoid(torch.matmul(input_data, w.double()))
+      obs = pyro.sample('obs', dist.Bernoulli(sigmoid), obs=observations)
+
+
+# Utility function to print latent sites' quantile information.
+def summary(samples):
+    site_stats = {}
+    for site_name, values in samples.items():
+        marginal_site = pd.DataFrame(values)
+        describe = marginal_site.describe(percentiles=[.05, 0.25, 0.5, 0.75, 0.95]).transpose()
+        site_stats[site_name] = describe[["mean", "std", "5%", "25%", "50%", "75%", "95%"]]
+    return site_stats
+
+def load_data():
+    npz_train_file = np.load('ds1.100_train.npz')
+    npz_test_file = np.load('ds1.100_test.npz')
+
+    X_train = torch.tensor(npz_train_file['X']).double()
+    y_train = torch.tensor(npz_train_file['y']).double()
+    y_train[y_train == -1] = 0
+    X_test = torch.tensor(npz_test_file['X']).double()
+    y_test = torch.tensor(npz_test_file['y']).double()
+    y_test[y_test == -1] = 0
+
+    return X_train, y_train, X_test, y_test
+
+
+def relbo(model, guide, *args, **kwargs):
```

Comment:
Is there an opportunity here to factor out relbo once and for all into its own file and then use it in a modular fashion?

Reply:
Yes, I think this should be possible and I got started on this. Only thing missing is to factor out the variable name that needs to be exposed when computing the trace of the approximation 
Will also be part cleaning up the BLR code.